feat!: multiple improvements to py_package and py_wheel

f0rmiga opened the pull request:
Each commit should be self-contained but at a high level:
- `py_package` now returns `PyInfo` with collected files and imports.
- `py_package` now has an `exclude` attr that's applied after the `packages`.
- Automatic third-party requirement collection in `py_wheel` via `auto_add_requirements`.
- Respect `imports` of `py_library` targets added to the `py_wheel`. This aligns with how Bazel assembles the PYTHONPATH under the runfiles.
- Added missing `imports` attributes to the `examples/wheel`.

It's only a breaking change for users not setting `imports` or some other behaviour that we cannot predict.
Despite it introducing a breaking change, the fixes users should apply are not complex and can be done by combining `imports` attributes and other features introduced by this PR.

AssemblyAI gently sponsored this work.

f0rmiga commented:
The current test failure is due to files clashing inside the action. I'll patch an alternative.

alexeagle commented:
Generally LGTM, it would be good to show an example of use and have a test that confirms that we build wheels with some dependencies automatically included.

rickeylev commented on examples/wheel/BUILD.bazel:
```diff
@@ -16,15 +16,28 @@ load("@bazel_skylib//rules:build_test.bzl", "build_test")
 
+compile_pip_requirements(
+    name = "requirements",
+    target_compatible_with = select({
```
I think you can simplify this to:
```
target_compatible_with = select({
  "@platforms//os:windows": ["@platforms//:incompatible"],
  "//conditions:default": []
})
```

f0rmiga replied:
This is the same comment as Alex. Addressed in 5747866.

rickeylev commented on WORKSPACE:
```diff
@@ -83,6 +83,16 @@ pip_parse(
+publish_install_deps()
+
+pip_parse(
+    name = "examples_wheel_deps",
```
Shouldn't this be in the example's workspace file...?

f0rmiga replied:
That example doesn't have its own WORKSPACE.

rickeylev commented on examples/wheel/lib/BUILD.bazel:
```diff
@@ -21,12 +21,31 @@ licenses(["notice"])  # Apache 2.0
+    name = "module_with_pip_deps",
+    srcs = ["module_with_pip_deps.py"],
+    imports = ["../../.."],
+    target_compatible_with = select({
```
Same comment here as the other TCW usage.

f0rmiga replied:
Addressed in dbb0710.

rickeylev commented on examples/wheel/requirements.in:
```diff
@@ -0,0 +1,2 @@
+Django
```
just making sure: this is "Django" with a capital D, not lowercase?

rickeylev commented on examples/wheel/wheel_test.py:
```diff
@@ -394,11 +395,39 @@ def test_rule_sets_stamped_version_in_wheel_metadata(self):
+                for line in fp:
+                    if line.startswith(b"Requires-Dist:"):
+                        requires.append(line.decode().strip())
+            self.assertEqual(4, len(requires))
```
nit: put the length check last

More preferable, assert against an expected list of all the values. This implicitly requires exactly equality, plus providers more context if there are errors.

f0rmiga replied:
Addressed in ca608f6.

rickeylev commented on examples/wheel/wheel_test.py:
```diff
@@ -394,11 +395,39 @@ def test_rule_sets_stamped_version_in_wheel_metadata(self):
+
+            requires = []
+            with zf.open(metadata_file) as fp:
+                for line in fp:
```
Is the order of entries guaranteed? If not, please sort or use a set or comparison.

f0rmiga replied:
It should be, but I sorted them anyways in e3767dd so that we don't have to care about this in the future in case the code that produces the wheel changes the behaviour.

rickeylev commented on python/private/py_package.bzl:
```diff
@@ -31,36 +31,62 @@ def _path_inside_wheel(input_file):
-        # TODO: flattening depset to list gives poor performance,
-        for input_file in inputs.to_list():
-            wheel_path = _path_inside_wheel(input_file)
-            for package in packages:
-                if wheel_path.startswith(package):
-                    filtered_files.append(input_file)
+            # TODO: flattening depset to list gives poor performance,
+            for input_file in input_files.to_list():
+                path_inside_wheel = _path_inside_wheel(input_file)
+                for package in packages:
+                    if path_inside_wheel.startswith(package) and not exclude:
+                        filtered_files.append(input_file)
+                        if path_inside_wheel.endswith(".py"):
+                            add_imports = True
+                    else:
+                        for excluded in exclude:
+                            if path_inside_wheel.startswith(package) and not path_inside_wheel.startswith(excluded):
+                                filtered_files.append(input_file)
+                                if path_inside_wheel.endswith(".py"):
+                                    add_imports = True
```
Oh wow this is really nested. Can you factor this out into some helper functions? This is 4 loops plus conditionals deep.

chrislovecnm replied:
And add some comments pls.

f0rmiga replied:
Now that you mention...
Addressed in 2bd26c6.

rickeylev commented on python/private/py_package.bzl:
```diff
@@ -31,36 +31,62 @@ def _path_inside_wheel(input_file):
-            for package in packages:
-                if wheel_path.startswith(package):
-                    filtered_files.append(input_file)
+            # TODO: flattening depset to list gives poor performance,
```
What's there TODO? It looks like the underlying objects being operated on are, rightly so, depsets; how would the flattening be removed?

f0rmiga replied:
This comment was here in the first place. I didn't bother removing it, and yes, there's no option.

rickeylev commented on python/private/py_package.bzl:
```diff
@@ -31,36 +31,62 @@ def _path_inside_wheel(input_file):
-    if not packages:
-        filtered_inputs = inputs
+    exclude = [p.replace(".", "/") for p in ctx.attr.exclude]
+    input_files = depset(transitive = [dep[PyInfo].transitive_sources for dep in ctx.attr.deps] + [dep[DefaultInfo].default_runfiles.files for dep in ctx.attr.deps])
```
Please reformat this line to make it shorter; it's almost two screen columns wide and hard to read in the diff.

f0rmiga replied:
Fixed in 910a2b0.

rickeylev commented on python/private/py_package.bzl:
```diff
@@ -31,36 +31,62 @@ def _path_inside_wheel(input_file):
-    if not packages:
-        filtered_inputs = inputs
+    exclude = [p.replace(".", "/") for p in ctx.attr.exclude]
+    input_files = depset(transitive = [dep[PyInfo].transitive_sources for dep in ctx.attr.deps] + [dep[DefaultInfo].default_runfiles.files for dep in ctx.attr.deps])
```
There's also .symlink, .root_symlinks, and .empty_filenames to consider when determining the set of files runfiles is going to materialize.

f0rmiga replied:
Addressed in 395a664.

rickeylev commented on python/private/py_wheel.bzl:
```diff
@@ -17,6 +17,62 @@
+    Args:
+        _: The bazel target the aspect is applied to.
+        ctx: The aspect context.
+    Returns: RequirementsInfo
```
nit: I'm surprised stardoc didn't complain about this. Well, only half surprised. Sometimes it demands "\nReturns:\n", and sometimes not, for reasons that aren't clear to me.

f0rmiga replied:
Addressed in 3295424. This was a carry-over, and it's not necessary. I never add stardoc to rule implementations.

rickeylev commented on python/private/py_wheel.bzl:
```diff
@@ -17,6 +17,62 @@
+    """
+
+    requirement = None
+    if hasattr(ctx.rule.attr, "tags"):
```
Is this guard here for a reason? tags is part of all rules, so this should always be true. If not, please add a comment about the edge case.

f0rmiga replied:
Addressed in 49710a0.

rickeylev commented on python/private/py_wheel.bzl:
```diff
@@ -17,6 +17,62 @@
+        pypi_version = None
+        for tag in ctx.rule.attr.tags:
+            if tag.startswith("pypi_name="):
+                pypi_name = tag[len("pypi_name="):]
```
Are these magic tags documented?

f0rmiga replied:
Nope, they are very magical.

rickeylev commented on python/private/py_wheel.bzl:
```diff
@@ -17,6 +17,62 @@
+        if pypi_name and pypi_version:
+            requirement = "{}=={}".format(pypi_name, pypi_version)
+
+    if hasattr(ctx.rule.attr, "deps"):
```
Similar comment to before: Add a comment explaining what is being visited where it's expected that it the rule won't have the attribute.

Same comment on the "srcs" check below, too

f0rmiga replied:
When traversing the dependency tree, srcs and deps are not attributes of every rule implementation. Nothing fancy going on here.

rickeylev commented on python/private/py_wheel.bzl:
```diff
@@ -352,6 +414,33 @@ def _py_wheel_impl(ctx):
+    filtered = []
+    for dep in deps:
+        files = dep[PyInfo].transitive_sources
+        imports = [imports.replace("\\", "/") for imports in dep[PyInfo].imports.to_list()]
```
Where is the \\ coming from? Windows? Please add a comment to that effect.

f0rmiga replied:
In more than 20 years in programming, I never saw a case where a backslash in a path meant something else other than dealing with Windows, and because it's so obvious, I won't add a comment.

rickeylev commented on python/private/py_wheel.bzl:
```diff
@@ -352,6 +414,33 @@ def _py_wheel_impl(ctx):
+                    args = actions.args()
+                    args.add("--source", file.path)
+                    args.add("--destination", link.path)
+                    actions.run(
```
Please set mnemonic and progress message on any actions that run

f0rmiga replied:
Addressed in 34a898b.

alexeagle commented on examples/wheel/BUILD.bazel:
```diff
@@ -16,15 +16,28 @@ load("@bazel_skylib//rules:build_test.bzl", "build_test")
 
+compile_pip_requirements(
+    name = "requirements",
+    target_compatible_with = select({
```
nit: can be a tad shorter like https://github.com/aspect-build/bazel-lib/blob/035473873a20135d24f7697426b613731cf1301f/lib/private/docs.bzl#L20

```
        "@platforms//os:windows": ["@platforms//:incompatible"],
        "//conditions:default": [],
```

rickeylev commented on python/private/py_wheel.bzl:
```diff
@@ -352,6 +414,33 @@ def _py_wheel_impl(ctx):
+                    if link_path.startswith("/"):
+                        link_path = link_path[1:]
+                    link_path = links_subdir + "/" + link_path
+                    link = actions.declare_directory(link_path) if file.is_directory else actions.declare_file(link_path)
```
This action can probably be replaced with just actions.symlink() when the file isn't a directory?

f0rmiga replied:
Trust me; I tried all the obvious solutions. CI would always fail for one of the platforms.

rickeylev commented on python/private/py_wheel.bzl:
```diff
@@ -352,6 +414,33 @@ def _py_wheel_impl(ctx):
+                    link_path = links_subdir + "/" + link_path
+                    link = actions.declare_directory(link_path) if file.is_directory else actions.declare_file(link_path)
+                    args = actions.args()
+                    args.add("--source", file.path)
```
nit: you can just pass "file". File objects are converted to their .path value when args expansion is done.

f0rmiga replied:
Only when it's not a directory (and here it can be a directory).

rickeylev commented on python/private/py_wheel.bzl:
```diff
@@ -352,6 +414,33 @@ def _py_wheel_impl(ctx):
+                    if link_path.startswith("/"):
+                        link_path = link_path[1:]
+                    link_path = links_subdir + "/" + link_path
+                    link = actions.declare_directory(link_path) if file.is_directory else actions.declare_file(link_path)
```
I'm surprised a directory is ended up as an input? how is that happening?

f0rmiga replied:
https://github.com/bazelbuild/rules_python/blob/0efcd94d0ee6e1e56b27d25469c28502282fab5b/examples/wheel/BUILD.bazel#L45-L49C2

alexeagle commented on python/private/py_wheel.bzl:
```diff
@@ -17,6 +17,62 @@
+                pypi_version = tag[len("pypi_version="):]
+
+        if pypi_name and pypi_version:
+            requirement = "{}=={}".format(pypi_name, pypi_version)
```
You already know: this is troubling, that ppl will publish libraries to pypi with exact version pinning like this rather than only a lower-bound. It will lead to incompatibilities downstream, when a constraint solver can't find a solution.

I suppose the alternative is for us to lookup the constraint from the `requirements.in` file (but could be a constraints.txt too).

Or maybe we should let users choose to use some semver range operator other than `==` in the thing we generate here?

f0rmiga replied:
Addressed in f2d7ae1. It can be a later work if we see the need to be fancier.

rickeylev commented on tools/linker.py:
```diff
@@ -0,0 +1,61 @@
+import shutil
+
+
+def create_link(source, destination, fallback_copy=False):
```
This seems like rather advanced logic just to create some symlinks. What edge case is it accounting for that I'm not seeing?

f0rmiga replied:
Handling of different platforms/setups. This is related to my previous response to your comment on using simpler logic for symlinking.

rickeylev commented on tools/linker.py:
```diff
@@ -0,0 +1,61 @@
+            )
+        else:
+            os.makedirs(os.path.dirname(destination), exist_ok=True)
+            (_copy if fallback_copy else _link)(source, destination)
```
This is clever, but not very readable. Just use a simple if-else, please.

f0rmiga replied:
Thanks for the compliment. Addressed in d4cbd53.

alexeagle commented on tools/linker.py:
```diff
@@ -0,0 +1,61 @@
+            (_copy if fallback_copy else _link)(source, destination)
+    except OSError as e:
+        if not fallback_copy:
+            return create_link(source, destination, fallback_copy=True)
```
that's surprising, for performance wouldn't we prefer a hardlink and only fallback to copying? @gregmagolan knows a lot about it from aspect_bazel_lib and rules_js

f0rmiga replied:
That's what the code is doing.

chrislovecnm commented on examples/wheel/lib/BUILD.bazel:
```diff
@@ -21,12 +21,31 @@ licenses(["notice"])  # Apache 2.0
 py_library(
     name = "simple_module",
     srcs = ["simple_module.py"],
```
Can we start leaving verbose comments in the examples?

f0rmiga replied:
Could help with those? I'm not the best person with the example comments.

chrislovecnm commented on examples/wheel/lib/module_with_pip_deps.py:
```diff
@@ -0,0 +1,19 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import django
```
Can we write a unit test for this please? The tests have helped me catch bugs with bazel.

chrislovecnm replied:
We may have a test. Correct me if I am wrong.

chrislovecnm reviewed (commented):
Thanks for all of the work. Mostly I have nits around documentation.

f0rmiga commented:
Thanks for the reviews, @alexeagle, @rickeylev and @chrislovecnm. I believe I addressed most of the comments. I only pushed back the ones I really thought didn't make sense.

f0rmiga commented on python/private/py_wheel.bzl:
```diff
@@ -436,6 +428,8 @@ def _inputs_to_package(actions, workspace_name, deps, links_subdir, linker):
                         arguments = [args],
                         executable = linker,
+                        progress_message = "Linking {} to {}".format(file.path, link.path),
+                        mnemonic = "Link",
```
Makes sense, thanks.

f0rmiga commented on python/private/py_wheel.bzl:
```diff
@@ -436,6 +428,8 @@ def _inputs_to_package(actions, workspace_name, deps, links_subdir, linker):
                         outputs = [link],
                         arguments = [args],
                         executable = linker,
+                        progress_message = "Linking {} to {}".format(file.path, link.path),
```
TIL

f0rmiga commented on tools/linker.py:
```diff
@@ -18,19 +18,20 @@
             os.makedirs(os.path.dirname(destination), exist_ok=True)
-            (_copy if fallback_copy else _link)(source, destination)
+            copy_function(source, destination)
     except OSError as e:
```
Because the filesystem might not accept symlinks and we have to fallback to copying.

f0rmiga commented on tools/linker.py:
```diff
@@ -18,19 +18,20 @@
 
 
 def create_link(source, destination, fallback_copy=False):
+    copy_function = _copy if fallback_copy else _link
```
Did I? I'm not worried about performance. I don't understand the question, or your concerns.

f0rmiga commented on python/private/py_wheel.bzl:
```diff
@@ -25,45 +25,37 @@ RequirementsInfo = provider(
             ],
         )
-    elif hasattr(ctx.rule.attr, "srcs"):
+    if hasattr(ctx.rule.attr, "srcs"):
```
Because srcs _may_ be plain source files, but it's not mandatory.

f0rmiga commented on python/private/py_package.bzl:
```diff
@@ -73,6 +76,46 @@ def _py_package_impl(ctx):
+            return True
+        for excluded in exclude:
+            if not path_inside_wheel.startswith(excluded):
+                return True
```
Thanks, I didn't know it.

f0rmiga commented on python/private/py_package.bzl:
```diff
@@ -35,32 +35,35 @@ def _py_package_impl(ctx):
+    ] + [
+        dep[DefaultInfo].default_runfiles.empty_filenames
+        for dep in ctx.attr.deps
+    ]
```
Done in d8ddc71.

rickeylev commented:
re: using ctx.actions.symlink instead of a custom program to perform essentially the same

> Trust me; I tried all the obvious solutions. CI would always fail for one of the platforms.

Well, still, would you mind elaborating a bit, for the rest of us, please?

ctx.actions.symlink should be platform agnostic, and should work with both files and directories. A specialized program to perform an otherwise standard operation isn't bad per-se, but it does mean the specialized program needs to explain what about the standard method is lacking that justifies needing a specialized program.

-----

re: why all the logic about symlink vs copying and error handling:

> Because the filesystem might not accept symlinks and we have to fallback to copying.

What filesystem or platform is that? For Windows, we basically require that symlinks are enabled now; I'm pretty sure there's a few other places that rely on symlink support now. Mac and Linux's support symlinks.

This is the sort of detail that belongs in a comment -- OSError is a generic class of errors. Is this due to some esoteric permission issue? Something to do with hardlinks? Some esoteric platform or system that doesn't accept symlinks? While you may have gone through a gauntlet and know the answer, other readers don't. All they see is error handling code for code that doesn't look like it should be raising an error.

-----

re: magic pypi tags:

> Nope, they are very magical.

Ok, so, we can't have magic tags that affect behavior without some sort of explanation about that. For example, I can trace through the code and can see how they are literally used, but that doesn't really provide an explanation. It just leaves me with several questions. For example:

* Why are there two tags? If both aren't present, they are ignored. The two tags are,
  in the end concatenated into a single value -- why are they separate tags?
* Why does the last value win? Is this intentional, or just accidental?
* Why can a target have only one requirement via its tags?
* Why are tags used instead an augmenting target in `data` and/or having RequirementsInfo provided directly?

I'm not saying we need a whole design doc or page long .md file about this. We do, however, need enough to understand the feature. That, and any publicly undocumented feature needs some sort of implementation comment about it, otherwise it's just a haunted graveyard people are afraid to touch.

-----

re: visiting srcs and gathering RequirementsInfo from them:

> Because srcs may be plain source files, but it's not mandatory.

Ok, but that still doesn't really answer the question: providing a requirement at an individual source-file level looks gratitious. A source file doesn't really do much in isolation; it's part of some unit (e.g. a py_library), which, as a unit, has requirements as to its needs.

chrislovecnm commented:
@f0rmiga ping
